### PR TITLE
Add SSE4.2 Path for CRC32, add A32 variant, add tests for non-castagnoli variants.

### PR DESCRIPTION
**Pull request by riperiperi**
This further improves the codegen for CRC32 instructions on A64, and adds it to A32. The tests that don't work in unicorn have been mirrored with a prebaked table of inputs and expected values, so they at least have some kind of testing going on.

The x86 CRC32 instruction, available since SSE4.2, allows quick calculation of crc using the castagnoli polynomial using just one instruction. We still need to use the Pclmulqdq or SoftFallback paths for non-castagnoli variants, but this is a nice boost for situations where it is used.

Tests are passing with these optimisations disabled. (SSE41, then Pclmulqdq)

The bits used for prefixes in the x86 assembler were changed to allow multiple to be present at once. They are checked in priority order - prefix 66 appears before f2, for example.

Games affected by addition of A32 instruction: Monster Hunter (all versions), Ni No Kuni (needs other instructions)

### CODE REVIEW

File: ARMeilleure/Instructions/InstEmitHashHelper.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            {
+                return size switch
+                {
+                    3 => EmitCrc32Optimized64(context, crc, value, castagnoli),
+                    _ => EmitCrc32Optimized(context, crc, value, castagnoli, size),
+                };
+            }
+            else
+            {
+                string name = (size, castagnoli) switch {
```
Comment (gdkchan) [inline ARMeilleure/Instructions/InstEmitHashHelper.cs:49]: nit: `{` on new line.

File: ARMeilleure/Instructions/InstEmitHashHelper.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            crc = context.VectorInsert(context.VectorZero(), crc, 0);
+
+            switch (size)
+            {
+                case 0: data = context.VectorInsert8(context.VectorZero(), data, 0); break;
+                case 1: data = context.VectorInsert16(context.VectorZero(), data, 0); break;
+                case 2: data = context.VectorInsert(context.VectorZero(), data, 0); break;
+            }
+
+            int bitsize = (8 << size);
```
Comment (gdkchan) [inline ARMeilleure/Instructions/InstEmitHashHelper.cs:79]: nit: Parenthesis is not necessary here.

File: Ryujinx.Tests/Cpu/CpuTest.cs [hunk 3]
```diff
@@ -167,41 +167,42 @@ protected void Opcode(uint opcode)
             }
         }
 
-        protected void ExecuteOpcodes()
+        protected void ExecuteOpcodes(bool runUnicorn = true)
```
Comment (gdkchan) [inline Ryujinx.Tests/Cpu/CpuTest.cs:170]: There are other instructions that are broken on Unicorn aswell (such as the AES instructions from what I recall, where we have pre-baked results), maybe we should do something similar for those aswell?

File: ARMeilleure/CodeGen/X86/PreAllocator.cs [hunk 4]
```diff
@@ -1294,11 +1294,22 @@ private static bool IsSameOperandDestSrc1(Operation operation)
                 case Instruction.VectorInsert16:
                 case Instruction.VectorInsert8:
                     return !HardwareCapabilities.SupportsVexEncoding;
+
+                case Instruction.Extended:
+                    return IsIntrinsicSameOperandDestSrc1(operation);
```
Comment (riperiperi) [inline ARMeilleure/CodeGen/X86/PreAllocator.cs:1299]: Needs to fall through to `IsVexSameOperandDestSrc1`

File: Ryujinx.Tests/Cpu/CpuTest.cs (hunk 3 above)
Comment (riperiperi) [reply Ryujinx.Tests/Cpu/CpuTest.cs:170]: AES was only broken on A32, the tests for A64 seem to call it without issue.

Comment (riperiperi) [reply Ryujinx.Tests/Cpu/CpuTest.cs:170]: (this was added to CpuTest32 first for that reason.)

Comment (gdkchan) [reply Ryujinx.Tests/Cpu/CpuTest.cs:170]: A64 AES tests also have pre-baked results: [https://github.com/Ryujinx/Ryujinx/blob/master/Ryujinx.Tests/Cpu/CpuTestSimdCrypto.cs#L18](https://github.com/Ryujinx/Ryujinx/blob/master/Ryujinx.Tests/Cpu/CpuTestSimdCrypto.cs#L18)

Not sure why it's comparing with Unicorn, but form what I recall it's broken.

**Comment by EmulationFanatic** [comment]
As mentioned, [Monster Hunter Generation Ultimate](https://github.com/Ryujinx/Ryujinx-Games-List/issues/511) should be fixed with this PR.

File: ARMeilleure/Instructions/InstEmitHashHelper.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+using static ARMeilleure.IntermediateRepresentation.OperandHelper;
+using static ARMeilleure.Instructions.InstEmitSimdHelper;
+
+namespace ARMeilleure.Instructions
+{
+    static class InstEmitHashHelper
+    {
+        public const uint Crc32RevPoly = 0xedb88320;
+        public const uint Crc32cRevPoly = 0x82f63b78;
```
Comment (LDj3SNuD) [inline ARMeilleure/Instructions/InstEmitHashHelper.cs:16]: They seem unused.

File: Ryujinx.Tests/Cpu/CpuTestAluBinary32.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
+﻿#define AluBinary32
+
+using ARMeilleure.State;
+using NUnit.Framework;
+using System;
+
+namespace Ryujinx.Tests.Cpu
+{
+
```
Comment (LDj3SNuD) [inline Ryujinx.Tests/Cpu/CpuTestAluBinary32.cs:9]: Remove empty line.

File: Ryujinx.Tests/Cpu/CpuTestAluBinary.cs [hunk 7]
```diff
@@ -8,8 +9,78 @@ namespace Ryujinx.Tests.Cpu
...
+                new CrcTest(0xffffffffu, 0x80_00_00_00_00_00_00_00u, false, 0x2dfd1072, 0xbe26ed00, 0xdebb20e3, 0x7765a3b6),
+                new CrcTest(0xffffffffu, 0xff_ff_ff_ff_ff_ff_ff_ffu, false, 0x00ffffff, 0x0000ffff, 0x00000000, 0xdebb20e3),
+                new CrcTest(0xffffffffu, 0xa0_02_f1_ca_52_78_8c_1cu, false, 0x39fc4c3d, 0xbc5f7f56, 0x4ed8e906, 0x12cb419c)
+            };
+        }
+#endregion
+
         private const int RndCnt = 2;
 
+        [Test, Combinatorial]
```
Comment (LDj3SNuD) [inline Ryujinx.Tests/Cpu/CpuTestAluBinary.cs:58]: `Combinatorial` is not necessary.

File: Ryujinx.Tests/Cpu/CpuTestAluBinary32.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+                new CrcTest32(0xffffffffu, 0x00_00_00_00u, true, 0xad82acae, 0x0e9e882d, 0xb798b438),
+                new CrcTest32(0xffffffffu, 0x7f_ff_ff_ffu, true, 0x00ffffff, 0x0000ffff, 0x82f63b78),
+                new CrcTest32(0xffffffffu, 0x80_00_00_00u, true, 0xad82acae, 0x0e9e882d, 0x356e8f40),
+                new CrcTest32(0xffffffffu, 0xff_ff_ff_ffu, true, 0x00ffffff, 0x0000ffff, 0x00000000),
+                new CrcTest32(0xffffffffu, 0x9d_cb_12_f0u, true, 0x5eecc3db, 0xbb6111cb, 0xcfb54fc9)
+            };
+        }
+#endregion
+
+        [Test, Combinatorial]
```
Comment (LDj3SNuD) [inline Ryujinx.Tests/Cpu/CpuTestAluBinary32.cs:69]: `Combinatorial` is not necessary.

File: Ryujinx.Tests/Cpu/CpuTestAluBinary.cs (hunk 7 above)
Comment (riperiperi) [reply Ryujinx.Tests/Cpu/CpuTestAluBinary.cs:58]: For future reference, is this implicit? Because combinatorial is the behaviour I'm going for.

Comment (LDj3SNuD) [reply Ryujinx.Tests/Cpu/CpuTestAluBinary.cs:58]: Yes: https://docs.nunit.org/articles/nunit/writing-tests/attributes/combinatorial.html
However, it should only be used when the total number of tests generated remains low, say in the order of hundreds.

**Pull request merged**